crates/astria-sequencer-relayer/src/relayer/celestia_client/mod.rs

```diff
@@ -245,6 +243,59 @@ impl CelestiaClient {
         trace!(?response);
         block_height_from_response(response)
     }
+
+    /// Repeatedly sends `GetTx` until a successful response is received.  Returns the height of the
+    /// Celestia block in which the blobs were submitted.
+    async fn confirm_submission(&mut self, tx_hash: TxHash) -> u64 {
+        // The min seconds to sleep after receiving a GetTx response and sending the next request.
+        const MIN_POLL_INTERVAL_SECS: u64 = 1;
+        // The max seconds to sleep after receiving a GetTx response and sending the next request.
+        const MAX_POLL_INTERVAL_SECS: u64 = 30;
+        // How long to wait after starting `confirm_submission` before starting to log errors.
+        const START_LOGGING_DELAY: Duration = Duration::from_secs(15);
+        // The minimum duration between logging errors.
+        const LOG_ERROR_INTERVAL: Duration = Duration::from_secs(5);
+
+        let start = Instant::now();
+        let mut logged_at = start;
+
+        let mut log_if_due = |maybe_error: Option<TrySubmitError>| {
+            if start.elapsed() <= START_LOGGING_DELAY || logged_at.elapsed() <= LOG_ERROR_INTERVAL {
+                return;
+            }
+            let reason = maybe_error.map_or("transaction still pending".to_string(), |error| {
+                astria_eyre::eyre::Report::new(error).to_string()
+            });
+            warn!(
+                reason,
+                tx_hash = tx_hash.0,
+                elapsed_seconds = start.elapsed().as_secs_f32(),
+                "waiting to confirm blob submission"
+            );
+            logged_at = Instant::now();
+        };
+
+        let mut sleep_secs = MIN_POLL_INTERVAL_SECS;
+        loop {
+            tokio::time::sleep(Duration::from_secs(sleep_secs)).await;
+            match self.get_tx(tx_hash.clone()).await {
+                Ok(Some(height)) => {
+                    // allow: the height is never negative.
+                    #[allow(clippy::cast_sign_loss)]
```

Comment:
Technically, CometBft/Cosmos heights *can* be negative because the intention is to be able to index from latest_height.

However, in the case here this would mean Celestia violated the (implicit?) contract of only reporting positive integers. I think I would prefer an explicit `try_into` here.

Reply:
> CometBft/Cosmos heights can be negative because the intention is to be able to index from latest_height.

Wow - that's extremely surprising to me!  I can't believe they'd call something like that a height.  To me, the height should be an immutable part of the block.

Anyway, thanks for the info.  I'll make the change as suggested.

Reply:
Done in [9fca78c](https://github.com/astriaorg/astria/pull/963/commits/9fca78c7b55dce3cefe5b5ce284fa4843ca30fd5).